kong/clustering/data_plane.lua

```diff
@@ -175,55 +170,46 @@ function _M:communicate(premature)
   --                 and is also responsible for handling timeout detection
 
   local ping_immediately
-  local config_exit
   local next_data
 
-  local config_thread = ngx.thread.spawn(function()
-    while not exiting() and not config_exit do
-      local ok, err = config_semaphore:wait(1)
-      if ok then
-        local data = next_data
-        if data then
-          local msg = assert(inflate_gzip(data))
-          yield()
-          msg = assert(cjson_decode(msg))
-          yield()
-
-          if msg.type == "reconfigure" then
-            if msg.timestamp then
-              ngx_log(ngx_DEBUG, _log_prefix, "received reconfigure frame from control plane with timestamp: ",
-                                 msg.timestamp, log_suffix)
-
-            else
-              ngx_log(ngx_DEBUG, _log_prefix, "received reconfigure frame from control plane", log_suffix)
-            end
-
-            local config_table = assert(msg.config_table)
-            local pok, res
-            pok, res, err = pcall(config_helper.update, self.declarative_config,
-                                  config_table, msg.config_hash, msg.hashes)
-            if pok then
-              if not res then
-                ngx_log(ngx_ERR, _log_prefix, "unable to update running config: ", err)
-              end
-
-              ping_immediately = true
-
-            else
-              ngx_log(ngx_ERR, _log_prefix, "unable to update running config: ", res)
-            end
-
-            if next_data == data then
-              next_data = nil
-            end
-          end
-        end
+  local on_recv_config = function(data)
+    if exiting() or not data then
+      return
+    end
 
-      elseif err ~= "timeout" then
-        ngx_log(ngx_ERR, _log_prefix, "semaphore wait error: ", err)
-      end
+    local msg = assert(inflate_gzip(data))
```

Comment:
Another question regarding this change:

The reason why @bungle previously introduced the `config_exit` variable with semaphore is to perform config loading coalescing, that is, if multiple config are received in short succession, only the last received config will be applied. With this change using worker events this seems to be no longer the case and could have performance penalties. Could you confirm?

Ping @bungle as well.

Reply:
I think it is a problem for this worker-events change.

Now in the master, the data plane will use the previously received config data and ignore the other received data until the reconfiguration is complete.

Reply:
The coalescing code:
https://github.com/Kong/kong/blob/master/kong/clustering/data_plane.lua#L269-L274

`config_exit` seems to be a way to notify `config_thread` about that either read or write thread was exited, so it has to exit too.

Reply:
Yep, that's precisely the problem @bungle tried to address in https://github.com/Kong/kong/pull/6299, so it might be inappropriate to use worker events to replace semaphore.

Reply:
I add a flag `is_reconfiguring`, which will have the same effect as semaphore.

Reply:
@chronolaw With this latest change the code seems to be more complex than before, do you think it is still a worthy refactor to do? We are forcing worker events to do something that it is not best at (coalescing) so we need another flag to emulate it. 

Say, the current PR actually adds more code than it removes now.

Reply:
I think the usage of three threads is confusing, worker-events is easier to understand than semaphore.

Reply:
Maybe, but after we have identified those issues (coalescing, bug in return value check), it seems that the extra is the most straightforward way to implement. Wouldn't adding another flag also contribute to confusion of the implementation in this case?

Especially, considering now we are removing 73 lines of code and adding back 81 lines, the benefit of the refactor appears to quickly diminish.

Reply:
@dndx , It looks reasonable. @bungle , what's your opinion?

If both of you think it is not worthy, I would close this PR.